Infrastructure.DataAccess/Migrations/SQLScripts/Deactive_Expired_Advice_And_Schedule_Missing_Jobs.sql

```diff
@@ -0,0 +1,217 @@
+		SELECT [Id]
+		FROM [Kitos].[dbo].[Advice]
+		WHERE Scheduling IS NOT NULL AND StopDate IS NOT NULL AND Scheduling != 0 AND StopDate < GETUTCDATE() AND IsActive = 1
+		AND [Id] NOT IN (
+			SELECT *
+			FROM @HFAdvice
+		)
+	
+		UPDATE [Kitos].[dbo].[Advice]
+		SET IsActive = 0
+		WHERE Id IN (
+			SELECT [AdviceId]
+			FROM @AdvicesToDisable
+		)
+	
+
+		/* Part 4: Locate active advice which have no related hangfire data */
+		DECLARE @AdvicesToAddToHF TABLE 
+		(
+			AdviceId int,
+			AlarmDate datetime2
+		)
+	
+		INSERT INTO @AdvicesToAddToHF
+		SELECT [Id], [AlarmDate]
+		FROM [Kitos].[dbo].[Advice]
+		WHERE Scheduling IS NOT NULL AND Scheduling != 0 AND (StopDate >= GETUTCDATE() OR StopDate IS NULL) AND IsActive = 1
+		AND [Id] NOT IN (
+			SELECT *
+			FROM @HFAdvice
+		)
+
+	
+		/* Part 5: Add to hangfire as scheduled jobs */
+		DECLARE @NewHFJobs TABLE
+		(
+			[JobId] int,
+			[AdviceId] int
+		)
+
+
+		INSERT INTO [kitos_HangfireDB].[HangFire].[Job] ([StateName], [InvocationData], [Arguments], [CreatedAt], [ExpireAt])
+		OUTPUT INSERTED.Id, REVERSE(SUBSTRING(REVERSE(SUBSTRING(INSERTED.Arguments, 3, 8)), 3, 8)) INTO @NewHFJobs ([JobId], [AdviceId])
+		SELECT
+			'Scheduled',
+			'{"Type":"Core.ApplicationServices.AdviceService, Core.ApplicationServices, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null","Method":"CreateOrUpdateJob","ParameterTypes":"[\"System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089\"]","Arguments":null}',
+			CONCAT('["', Advices.[AdviceId], '"]'),
+			SYSUTCDATETIME(),
+			DATEADD(month, 1, SYSUTCDATETIME())
+		FROM @AdvicesToAddToHF AS Advices
+
+		INSERT INTO [kitos_HangfireDB].[HangFire].[JobParameter] ([JobId], [Name], [Value])
+		SELECT 
+			[JobId],
+			'CurrentCulture',
+			'"en-DK"'
+		FROM @NewHFJobs
+
+		INSERT INTO [kitos_HangfireDB].[HangFire].[JobParameter] ([JobId], [Name], [Value])
+		SELECT 
+			[JobId],
+			'CurrentUICulture',
+			'"en-US"'
+		FROM @NewHFJobs
+
+		INSERT INTO [kitos_HangfireDB].[HangFire].[State] ([JobId], [Name], [Reason], [CreatedAt], [Data])
+		SELECT 
+			NewJobs.[JobId],
+			'Scheduled', 
+			NULL,
+			SYSUTCDATETIME(), 
+			CONCAT('{"EnqueueAt":"', CONVERT(date, Advices.AlarmDate), 'T', CONVERT(time, Advices.AlarmDate), 'Z","ScheduledAt":"', CONVERT(date, SYSUTCDATETIME()), 'T', CONVERT(time, SYSUTCDATETIME()), 'Z', '"}')
+		FROM @NewHFJobs AS NewJobs
+			INNER JOIN
+			@AdvicesToAddToHF AS Advices ON NewJobs.AdviceId = Advices.AdviceId
+
+		UPDATE [kitos_HangfireDB].[HangFire].[Job]
+		SET [Job].[StateId] = [State].[Id]
+		FROM 
+			[kitos_HangfireDB].[HangFire].[State] AS [State]
+			INNER JOIN 
+			[kitos_HangfireDB].[HangFire].[Job] AS [Job]
+			ON [Job].Id = [State].JobId
+
+		UPDATE [kitos_HangfireDB].[HangFire].[Job]
+		SET [ExpireAt] = NULL
+		WHERE [Id] IN (
+			SELECT [JobId]
+			FROM @NewHFJobs
+		)
+
+		DECLARE @JobsWhichHasPassedAlarmDate TABLE
+		(
+			[JobId] int
+		)
+
+		INSERT INTO @JobsWhichHasPassedAlarmDate
```

Comment:
jeg kom lige til at tænke på om vi overhovedet behøves to sæt... hvis man nu sætter en alarmdata i fortiden, vil den så ikke bare blive afviklet? Det ville da rense scriptet en del for så vile der kun være een change table med udformning som JobsHasYetToPassAlarmDate... når du laver din merge insert, så laver du jo faktisk en "fortids-time"
![image](https://user-images.githubusercontent.com/13116889/134845933-b0d96c6c-ab4a-4563-9e87-9ed71774e047.png)
Når ovenstående læses af Hangfire, så er tide jo alligevel passeret og så er det jo lige meget om det er et år eller to sekunder. Det ville i hvert fald spare en masse script plads, for hvis det bare handler om "passeret" / "ikke passeret" for den der score, jamen så gør det jo vores liv langt nemmere

Reply:
Jeg får lige testet om de så kører med det samme eller om deres "Score" så er for lav til at hangfire aktiverer dem.

Reply:
Det virker. De bliver kørt med samme 👍

Reply:
Fedt jamen så sparer vi da nogle linjer ;-)